Add minimalistic implementation of subgraph matcher.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19322 Add minimalistic implementation of subgraph matcher.**

Differential Revision: [D14962182](https://our.internmc.facebook.com/intern/diff/D14962182)

Commit:
Add minimalistic implementation of subgraph matcher.

ZolotukhinM commented:
We can later add more features to this (e.g. nice iterators as in #17100). I'm putting this version for review now to unblock some of other work items.

bwasti commented on torch/csrc/jit/subgraph_matcher.cpp:
```diff
@@ -0,0 +1,200 @@
+}
+
+/**
+ * Recursively try to match pattern with the actual graph starting from the
```
is this recursive?

ZolotukhinM replied:
> is this recursive?

Yes, `matchNodes` call `matchValues`, and `matchValues` call `matchNodes`. The recursion stops when we reach `Param` nodes in the pattern or if we hit an already visited node/value.

bwasti commented on torch/csrc/jit/subgraph_matcher.cpp:
```diff
@@ -0,0 +1,200 @@
+  while (!blocks_to_visit.empty()) {
+    const Block* block = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (const Node* n : block->nodes()) {
```
maybe call out that this is a match on the imperative interpretation of the IR, and not the data flow graph

diamonds declared in different orders won't match, right?

ZolotukhinM replied:
Hm, they should match whatever the order is (I added a test for this). We do follow data flow graph, not the sequential list of nodes: `matchNodes` call `matchValues` for nodes' inputs, and `matchValues` proceed to values' defs calling `matchNodes` for them.

PS: This implementation lacks aliasing checks, which are required if we're going to replace matched subgraphs with something else - I don't forget about it, just plan to follow up on that separately.

bwasti reviewed (approved):
looks very good -- well tested

Commit:
Update on "Add minimalistic implementation of subgraph matcher."

Add minimalistic implementation of subgraph matcher.

gh-metadata: pytorch pytorch 19322 gh/ZolotukhinM/10/head

ZolotukhinM reviewed (commented):
Thanks for taking a look! I replied inline.

Commit:
Update on "Add minimalistic implementation of subgraph matcher."

Add minimalistic implementation of subgraph matcher.

gh-metadata: pytorch pytorch 19322 gh/ZolotukhinM/10/head

Commit:
Update on "Add minimalistic implementation of subgraph matcher."

Add minimalistic implementation of subgraph matcher.

gh-metadata: pytorch pytorch 19322 gh/ZolotukhinM/10/head

Commit:
Update on "Add minimalistic implementation of subgraph matcher."

Add minimalistic implementation of subgraph matcher.

gh-metadata: pytorch pytorch 19322 gh/ZolotukhinM/10/head
